facilitator/src/bin/facilitator.rs

```diff
@@ -739,30 +807,86 @@ fn intake_batch(sub_matches: &ArgMatches) -> Result<(), anyhow::Error> {
         is_first_from_arg(sub_matches),
     )?;
 
+    batch_intaker.generate_validation_share()
+}
+
+fn intake_batch_subcommand(sub_matches: &ArgMatches) -> Result<(), anyhow::Error> {
+    intake_batch(
+        sub_matches.value_of("aggregation-id").unwrap(),
+        sub_matches.value_of("batch-id").unwrap(),
+        sub_matches.value_of("date").unwrap(),
+        sub_matches,
+    )
+}
+
+fn intake_batch_worker(sub_matches: &ArgMatches) -> Result<(), anyhow::Error> {
     let intake_started: Counter = register_counter!(
         "intake_jobs_begun",
         "Number of intake-batch jobs that started (on the facilitator side)"
     )
-    .unwrap();
-    intake_started.inc();
-
-    let result = batch_intaker.generate_validation_share();
+    .context("failed to register metrics counter for started intakes")?;
 
     let intake_finished = register_counter_vec!(
         "intake_jobs_finished",
         "Number of intake-batch jobs that finished (on the facilitator side)",
         &["status"]
     )
-    .unwrap();
-    match result {
-        Ok(_) => intake_finished.with_label_values(&["success"]).inc(),
-        Err(_) => intake_finished.with_label_values(&["error"]).inc(),
+    .context("failed to register metrics counter for finished intakes")?;
+
+    let mut queue = intake_task_queue_from_args(sub_matches)?;
+
+    loop {
+        if let Some(task_handle) = queue.dequeue()? {
+            info!("dequeued task: {}", task_handle);
+            intake_started.inc();
+            let result = intake_batch(
+                &task_handle.task.aggregation_id,
+                &task_handle.task.batch_id,
+                &task_handle.task.date,
+                sub_matches,
+            );
+
+            match result {
+                Ok(_) => {
+                    intake_finished.with_label_values(&["success"]).inc();
+                    queue.acknowledge_task(task_handle)?;
+                }
+                Err(err) => {
+                    error!("error while processing task {}: {:?}", task_handle, err);
+                    intake_finished.with_label_values(&["error"]).inc();
```

Comment:
PubSub will delete a message if it is not acked after 7 days, and will try to redeliver the message with exponential backoff until then. I like the idea of relying on PubSub for retries -- it saves me the trouble of writing retry code. I have not set up any dead letter queue because I don't know what I would do with it, to be honest. Email someone? Alert about the job we're giving up on?

We could nack messages much more aggressively then 7 days. There's no point to retrying an intake job if the aggregation window it corresponds to is long gone, for instance. I think we could achieve that by looking at the message redelivery count in the PubSub API response, but I'd like to leave that for a future change: #295

Reply:
I'm reading up on this more and it seems like dead letter topics and more aggressive nacks are (1) a good idea and (2) pretty simple to wire up so I'll just do them in this PR.

Reply:
So I think I understand all this better now. We now create a dead letter topic for each subscription, which causes PubSub to [redeliver messages up to 5 times](https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions#deadletterpolicy) until they are ACKed. The [default retry policy](https://cloud.google.com/pubsub/docs/reference/rest/v1/projects.subscriptions#retrypolicy) means PubSub will use exponential backoff to wait between 10 seconds and 10 minutes to redeliver unacked messages. So that means the default retention policy of 7 days should be irrelevant -- six delivery attempts (1 + 5 redeliveries) should have gone by long before then, and messages that fail to be processed six times go to the dead letter queue for later examination by humans.

I'd also misunderstood how nacking works -- I thought a nack would make PubSub forward a message to the dead letter queue, but instead of short circuits the visibility timeout and allows a message to be redelivered sooner.

Reply:
Correct nack is a shortcut to have it redelivered and retried without having to wait for the Ack timeout. The 7 days retention is if you have replay enabled or if your workers are so busy that they don't get to it earlier. So there is still a value of having a retention policy.